File: internal/apiv2/admin_autoscan_scans.go
package apiv2

import (
	"context"
	"encoding/json"
	"errors"
	"strconv"
	"strings"

	"github.com/Silo-Server/silo-server/internal/api/handlers"
	"github.com/Silo-Server/silo-server/internal/autoscan"
	"github.com/Silo-Server/silo-server/internal/scanqueue"
	"github.com/danielgtaylor/huma/v2"
)

type AdminAutoscanScansService interface {
	ReadAdminAutoscanScans(context.Context, autoscan.ScanListFilter) ([]autoscan.ScanWithEvent, int, error)
}
type AdminAutoscanScanStatus string

func (AdminAutoscanScanStatus) Schema(huma.Registry) *huma.Schema {
	return &huma.Schema{Type: huma.TypeString, Enum: []any{"", scanqueue.StatusAccepted, scanqueue.StatusRunning, scanqueue.StatusCompleted, scanqueue.StatusFailed, scanqueue.StatusCancelled}}
}

const autoscanScanIDDescending = "id_desc"

type AdminAutoscanScansInput struct {
	Limit  int                     `query:"limit" default:"50" minimum:"1" maximum:"200"`
	Cursor string                  `query:"cursor" maxLength:"8192"`
	Status AdminAutoscanScanStatus `query:"status"`
	Search string                  `query:"q" maxLength:"1024"`
}
type AdminAutoscanScansOutput struct{ Body AdminAutoscanScansPage }
type AdminAutoscanScansPage struct {
	Items []AdminAutoscanScan `json:"items"`
	Page  *PageInfo           `json:"page"`
	Total int                 `json:"total" minimum:"0" doc:"Separate live count; not a snapshot of the page."`
}
type AdminAutoscanScan struct {
	ID               string   `json:"id"`
	LibraryID        string   `json:"library_id"`
	Mode             string   `json:"mode"`
	Path             string   `json:"path,omitempty"`
	Trigger          string   `json:"trigger"`
	Status           string   `json:"status"`
	ErrorMessage     string   `json:"error_message,omitempty"`
	RequestedAt      *Instant `json:"requested_at,omitempty"`
	StartedAt        *Instant `json:"started_at,omitempty"`
	CompletedAt      *Instant `json:"completed_at,omitempty"`
	AutoscanEventID  *string  `json:"autoscan_event_id,omitempty"`
	SourceID         *string  `json:"source_id,omitempty"`
	PluginID         string   `json:"plugin_id,omitempty"`
	CapabilityID     string   `json:"capability_id,omitempty"`
	EventStatus      string   `json:"event_status,omitempty"`
	EventCompletedAt *Instant `json:"event_completed_at,omitempty"`
}

func registerAdminAutoscanScans(reg *Registry) {
	cursors := NewCursors(reg.deps.CursorSecret)
	op := Operation{Operation: humaOp("GET", Prefix+"/admin/autoscan/scans", "listAdminAutoscanScans", "admin-autoscan", "Read bounded SQL offset pages with a separate live count. Mutable lifecycle timestamp descending/id descending order; signed continuation is not a snapshot or keyset guarantee."), Class: ClassActingAdmin, ServiceBacked: true}
	Register(reg, op, func(ctx context.Context, in *AdminAutoscanScansInput) (*AdminAutoscanScansOutput, error) {
		if reg.deps.AdminAutoscanScans == nil {
			return nil, unavailable("autoscan scan history")
		}
		filter := autoscan.ScanListFilter{Limit: in.Limit, Status: string(in.Status), Search: strings.TrimSpace(in.Search)}
		encoded, _ := json.Marshal([]any{filter.Status, filter.Search, filter.Limit})
		scope := CursorScope{OperationID: "listAdminAutoscanScans", Security: strconv.Itoa(claimsFrom(ctx).UserID) + "/" + profileFrom(ctx) + "/" + viewerScopeDigest(ctx), Filter: string(encoded), Sort: "lifecycle_desc", Tiebreaker: autoscanScanIDDescending}
		var pos struct{ Offset int }
		if in.Cursor != "" {
			if problem := cursors.Decode(scope, in.Cursor, &pos); problem != nil {
				return nil, problem
			}
		}
		if pos.Offset < 0 || pos.Offset > 1000000000 {
			return nil, NewProblem(TypeInvalidCursor, "Invalid scan history position.")
		}
		filter.Offset = pos.Offset
		rows, total, err := reg.deps.AdminAutoscanScans.ReadAdminAutoscanScans(ctx, filter)
		if errors.Is(err, handlers.ErrAdminAutoscanScansUnavailable) {
			return nil, unavailable("autoscan scan history")
		}
		if err != nil {
			return nil, serviceProblem(err)
		}
		if total < 0 || len(rows) > in.Limit {
			return nil, NewProblem(TypeInternalError, "Invalid scan history page.")
		}
		items := make([]AdminAutoscanScan, 0, len(rows))
		for _, r := range rows {
			if r.ID == "" || r.MediaFolderID <= 0 {
				return nil, NewProblem(TypeInternalError, "Invalid scan history identity.")
			}
			row := AdminAutoscanScan{ID: r.ID, LibraryID: strconv.Itoa(r.MediaFolderID), Mode: r.Mode, Path: r.Path, Trigger: r.Trigger, Status: r.Status, ErrorMessage: r.ErrorMessage, RequestedAt: instantPtr(r.RequestedAt), StartedAt: instantPtr(r.StartedAt), CompletedAt: instantPtr(r.CompletedAt), SourceID: r.SourceID, PluginID: r.PluginID, CapabilityID: r.CapabilityID, EventStatus: string(r.EventStatus), EventCompletedAt: instantPtr(r.EventCompletedAt)}
			if r.AutoscanEventID != nil {
				row.AutoscanEventID = new(strconv.FormatInt(*r.AutoscanEventID, 10))
			}
			items = append(items, row)
		}
		next := ""
		if len(rows) == in.Limit {
			pos.Offset += len(rows)
			next, err = cursors.Encode(scope, pos)
			if err != nil {
				return nil, serviceProblem(err)
			}
		}
		return &AdminAutoscanScansOutput{Body: AdminAutoscanScansPage{Items: items, Page: Paginated(items, next).Page, Total: total}}, nil
	})
}

File: web/src/api/v2/adminAutoscanScans.ts
import {
  isCapturedProfileAuthorityActive,
  StaleApiRequestContextError,
  type ProfileRequestContextSnapshot,
} from "@/api/client";
import type { AutoscanScan, AutoscanScanStatus, AutoscanEventStatus } from "@/api/types";
import { v2 } from "./request";
export type AutoscanScanQuery = {
  status?: AutoscanScanStatus;
  query?: string;
  limit?: number;
  offset?: number;
  enabled?: boolean;
};
function safeID(value: string): number {
  const n = Number(value);
  if (!/^[1-9][0-9]*$/.test(value) || !Number.isSafeInteger(n))
    throw new Error("Invalid scan history identity.");
  return n;
}
export async function readAdminAutoscanScans(
  profileContext: ProfileRequestContextSnapshot,
  params: AutoscanScanQuery,
): Promise<{ rows: AutoscanScan[]; total: number }> {
  const limit = params.limit ?? 50;
  const offset = params.offset ?? 0;
  if (
    !Number.isInteger(limit) ||
    limit < 1 ||
    limit > 200 ||
    !Number.isInteger(offset) ||
    offset < 0 ||
    offset % limit !== 0 ||
    offset / limit >= 100
  )
    throw new Error("Scan history page is outside the supported range.");
  let cursor: string | undefined;
  const seen = new Set<string>();
  for (let page = 0; page <= offset / limit; page++) {
    if (!isCapturedProfileAuthorityActive(profileContext)) throw new StaleApiRequestContextError();
    const result = await v2("GET /api/v2/admin/autoscan/scans", {
      profileContext,
      query: { limit, cursor, status: params.status, q: params.query },
    });
    if (!isCapturedProfileAuthorityActive(profileContext)) throw new StaleApiRequestContextError();
    if (
      !result.page ||
      typeof result.page.has_more !== "boolean" ||
      !Number.isSafeInteger(result.total) ||
      result.total < 0 ||
      result.items.length > limit
    )
      throw new Error("Invalid scan history page.");
    const next = result.page.next_cursor;
    if (result.page.has_more && (!next || seen.has(next)))
      throw new Error("Invalid scan history continuation.");
    if (page === offset / limit) {
      const ids = new Set<string>();
      const rows: AutoscanScan[] = result.items.map((row) => {
        if (!row.id || ids.has(row.id)) throw new Error("Invalid scan history identity.");
        ids.add(row.id);
        if (
          !["library", "subtree", "file"].includes(row.mode) ||
          !["accepted", "running", "completed", "failed", "cancelled"].includes(row.status)
        )
          throw new Error("Unsupported scan history state.");
        if (
          row.event_status &&
          !["running", "success", "error", "unresolved"].includes(row.event_status)
        )
          throw new Error("Unsupported scan event state.");
        return {
          ...row,
          library_id: safeID(row.library_id),
          autoscan_event_id:
            row.autoscan_event_id == null ? undefined : safeID(row.autoscan_event_id),
          mode: row.mode as AutoscanScan["mode"],
          status: row.status as AutoscanScanStatus,
          event_status: row.event_status as AutoscanEventStatus | undefined,
          source_id: row.source_id ?? undefined,
        };
      });
      return { rows, total: result.total };
    }
    if (!result.page.has_more) return { rows: [], total: result.total };
    seen.add(next!);
    cursor = next;
  }
  throw new Error("Scan history page unavailable.");
}
